cardano-cli/src/Cardano/CLI/Shelley/Run/SingleAddressWallet.hs

```diff
@@ -0,0 +1,163 @@
+
+import           Ouroboros.Network.Block (getTipPoint, pointSlot)
+
+import qualified Shelley.Spec.Ledger.Coin as Shelley (Coin (..))
+import qualified Shelley.Spec.Ledger.TxData as Shelley (Addr (..), TxId (..), TxIn (..), TxOut (..))
+import qualified Shelley.Spec.Ledger.UTxO as Shelley (UTxO (..))
+
+-- | Build a basic transaction that consists of a single input and output
+-- and covers its own fees.
+--
+-- For the input, we query the node's local state for a UTxO that is
+-- sufficient to cover the min fee of the transaction that will be
+-- constructed. For the output, we simply spend the change back to the sender.
+buildSingleAddressWalletTransaction
+  :: CodecConfig (ShelleyBlock TPraosStandardCrypto)
+  -> Network
+  -> SocketPath
+  -> Maybe SlotNo
+  -> SigningKey
+  -> ExceptT CliError IO TxSigned
+buildSingleAddressWalletTransaction _ _ _ _ (SigningKeyByron _) =
+  panic "buildSingleAddressWalletTransaction: SigningKeyByron"
+buildSingleAddressWalletTransaction codecCfg network sockPath mbTxTtl
+                                    skey@(SigningKeyShelley shelleySkey) = do
+    -- Query the local tip from chain sync protocol
+    tipPoint <- withIOManagerE $ \iocp ->
+      liftIO $ getTipPoint <$> getLocalTip iocp codecCfg network sockPath
+
+    -- Query the PParams from the node's local state
+    pparams <- firstExceptT NodeLocalStateQueryError $
+      queryPParamsFromLocalState network sockPath tipPoint
+
+    -- Build a transaction containing one dummy input (representing an input
+    -- that can cover the fee) and one dummy output (representing the change
+    -- output).
+    --
+    -- We then use this dummy transaction to calculate the min fee which we'll
+    -- utilize when constructing the real transaction.
+    let txTtl = chooseTxTtl tipPoint
+        dummyTxForFeeCalc = buildDummyShelleyTx txTtl
+        minFee = calculateShelleyMinFee pparams dummyTxForFeeCalc
+
+    -- Query the UTxO from the node's local state
+    utxo <- firstExceptT NodeLocalStateQueryError $
+      queryFilteredUTxOFromLocalState network sockPath (Set.singleton addr) tipPoint
+
+    -- Find an unspent transaction output that can be used to cover the fee.
+    (txIn, TxOut _ outAmount) <- hoistMaybe SingleAddressWalletError $
+      findUTxOToCoverFee utxo minFee
+
+    -- Construct a transaction output to send the change back to the origin address
+    let txOut = TxOut addr $ Lovelace (unLoveLace outAmount - unLoveLace minFee)
+
+    -- Construct and return the legitimate transaction.
+    pure $ buildAndSignShelleyTx
+      [txIn]
+      [txOut]
+      txTtl
+      minFee
+      network
+      [skey]
+  where
+    numDummyTxIns = 1
+    numDummyTxOuts = 1
+    vkey = PaymentVerificationKeyShelley $ deriveVerKey shelleySkey
+    addr = shelleyVerificationKeyAddress vkey Nothing
+
+    -- If the transaction TTL was provided, use that. Otherwise, use
+    -- @tipPointSlotNo + 100@.
```

Comment:
There's nothing significant about the number `100`. I just randomly picked it for now.